docs/source/developer/patterns.rst

```diff
@@ -168,6 +168,55 @@ that do not support them. When handling a resolved or rejected Promise,
 make sure to check for the current state (typically by checking an
 ``.isDisposed`` property) before proceeding.
 
+Server Requests
+---------------
+
+To allow hot-swapping the Jupyter Server in custom applications
+based of JupyterLab components, the request URL should be composed using
+the base URL from the ``ServerConnection.ISettings`` instance derived
+from the service manager, passed down via class constructors as needed.
+The `PageConfig.getBaseUrl()` should not be used directly.
```

Comment:
Once https://github.com/jupyterlab/jupyterlab/pull/17953 is merged I will apply:


```suggestion
the base URL from the :ts:interface:`services.ServerConnection.ISettings` instance derived
from the service manager, passed down via class constructors as needed.
The :ts:function:`coreutils.PageConfig.getBaseUrl()` should not be used directly.
```